### ISSUE TITLE
Level 4 warnings when building with Visual Studio 2017 with 2015 XP toolset

### ISSUE THREAD

**Issue by JoelTroch** [gabime/spdlog#761]
Building spdlog through CMake using Visual Studio 2017 with the 2015 XP toolset (`v140_xp`) and setting the warning level to 4 (`/W4`) will provoke the following warnings:

Commit: https://github.com/gabime/spdlog/commit/ddb3002bc16a68148d55b0280c4ce843ea415969

**C4127 - Conditional expression is a constant**
Concerned file: `format.h`.
Concerned line: 3538

```cpp
basic_memory_buffer<char_type> buffer;
if (FMT_USE_GRISU && sizeof(T) <= sizeof(double) && std::numeric_limits<double>::is_iec559) // C4127
{
    internal::fp fp_value(static_cast<double>(value));
    fp_value.normalize();

    // ...
```

**C4244 - Argument conversion from 'int' to 'const char', possible data loss**
Concerned file: `fmt_helper.h`.
Concerned lines: 54, 61, 79, 87.

Code:

```cpp
inline void pad2(int n, fmt::basic_memory_buffer<char, Buffer_Size> &dest)
{
    if (n > 99)
    {
        append_int(n, dest);
        return;
    }
    if (n > 9) // 10-99
    {
        dest.push_back('0' + (n / 10)); // C4244
        dest.push_back('0' + (n % 10));
        return;
    }
    if (n >= 0) // 0-9
    {
        dest.push_back('0');
        dest.push_back('0' + n); // C4244
        return;
    }
    // negatives (unlikely, but just in case, let fmt deal with it)
    fmt::format_to(dest, "{:02}", n);
}

template<size_t Buffer_Size>
inline void pad3(int n, fmt::basic_memory_buffer<char, Buffer_Size> &dest)
{
    if (n > 99)
    {
        append_int(n, dest);
        return;
    }
    if (n > 9) // 10-99
    {
        dest.push_back('0');
        dest.push_back('0' + n / 10); // C4244
        dest.push_back('0' + n % 10);
        return;
    }
    if (n >= 0)
    {
        dest.push_back('0');
        dest.push_back('0');
        dest.push_back('0' + n); // C4244
        return;
    }
    // negatives (unlikely, but just in case let fmt deal with it)
    fmt::format_to(dest, "{:03}", n);
}
```

**C4702 - Unreachable code**
Concerned file: `thread_pool.h`
Concerned line: 213

Code:
```cpp
// process next message in the queue
// return true if this thread should still be active (while no terminate msg was received)
bool process_next_msg_()
{
    async_msg incoming_async_msg;
    bool dequeued = q_.dequeue_for(incoming_async_msg, std::chrono::seconds(10));
    if (!dequeued)
    {
        return true;
    }

    switch (incoming_async_msg.msg_type)
    {
    case async_msg_type::flush:
    {
        incoming_async_msg.worker_ptr->backend_flush_();
        return true;
    }

    case async_msg_type::terminate:
    {
        return false;
    }

    default:
    {
        log_msg msg;
        incoming_async_msg.to_log_msg(msg);
        incoming_async_msg.worker_ptr->backend_log_(msg);
        return true;
    }
    }
    assert(false); // C4702
    return true; // should not be reached
}
```

**Comment by gabime** [comment]
Fixed in b0059b290f3b6e0697c2a207096b8a48e7e16832 (except for "the C4127 - Conditional expression is a constant" in format.h - it is part of fmtlib - please open a ticket about it in its [repo](https://github.com/fmtlib/fmt))

**Comment by JoelTroch** [comment]
Thanks a lot!
